File: Python3.x/19-Remove_Nth_Node_From_End_of_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        tmp = ListNode(None)
        tmp.next = head
        start, end = tmp, tmp
        for _ in range(n): start = start.next
        while start.next:
            start, end = start.next, end.next
        end.next = end.next.next
        return tmp.next


'''
加一个头结点，并使用双指针start和end。
start先向前移动n个节点，然后start和end同时移动，当start.next==None时，此时end.next指的就是需要删除的节点。
'''

File: C++/19-Remove Nth Node From End of List.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Remove Nth Node From End of List.
// Memory Usage: 8.4 MB, less than 89.92% of C++ online submissions for Remove Nth Node From End of List.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode res(NULL);
        res.next = head;
        ListNode* p1 = &res;
        ListNode* p2 = &res;
        for (int i = 0; i < n; ++i)
            p2 = p2->next;
        while (p2->next) {
            p1 = p1->next;
            p2 = p2->next;
        }
        p1->next = p1->next->next;
        return res.next;
    }
};